clippy_lints/src/manual_let_else.rs

```diff
@@ -0,0 +1,269 @@
+    restriction,
+    "manual implementation of a let else statement"
+}
+
+pub struct ManualLetElse {
+    msrv: Option<RustcVersion>,
+}
+
+impl ManualLetElse {
+    #[must_use]
+    pub fn new(msrv: Option<RustcVersion>) -> Self {
+        Self { msrv }
+    }
+}
+
+impl_lint_pass!(ManualLetElse => [MANUAL_LET_ELSE]);
+
+impl<'tcx> LateLintPass<'tcx> for ManualLetElse {
+    fn check_stmt(&mut self, cx: &LateContext<'_>, stmt: &'tcx Stmt<'tcx>) {
+        let if_let_or_match = if_chain! {
+            if meets_msrv(self.msrv, msrvs::LET_ELSE);
+            if !in_external_macro(cx.sess(), stmt.span);
+            if let StmtKind::Local(local) = stmt.kind;
+            if let Some(init) = local.init;
+            if !from_different_macros(init.span, stmt.span);
+            if let Some(if_let_or_match) = IfLetOrMatch::parse(cx, init);
+            then {
+                if_let_or_match
+            } else {
+                return;
+            }
+        };
+
+        match if_let_or_match {
+            IfLetOrMatch::IfLet(_let_expr, let_pat, if_then, if_else) => if_chain! {
+                if expr_is_simple_identity(let_pat, if_then);
+                if let Some(if_else) = if_else;
+                if expr_diverges(cx, if_else);
+                then {
+                    span_lint(
+                        cx,
+                        MANUAL_LET_ELSE,
+                        stmt.span,
+                        "this could be rewritten as `let else`",
+                    );
+                }
+            },
+            IfLetOrMatch::Match(_match_expr, arms, source) => {
+                if source != MatchSource::Normal {
+                    return;
+                }
+                // Any other number than two arms doesn't (neccessarily)
+                // have a trivial mapping to let else.
+                if arms.len() != 2 {
+                    return;
+                }
+                // We iterate over both arms, trying to find one that is an identity,
+                // one that diverges. Our check needs to work regardless of the order
+                // of both arms.
+                let mut found_identity_arm = false;
+                let mut found_diverging_arm = false;
+                for arm in arms {
+                    // Guards don't give us an easy mapping to let else
+                    if arm.guard.is_some() {
+                        return;
+                    }
+                    if expr_is_simple_identity(arm.pat, arm.body) {
+                        found_identity_arm = true;
+                    } else if expr_diverges(cx, arm.body) && pat_allowed_for_else(cx, arm.pat) {
+                        found_diverging_arm = true;
+                    }
+                }
+                if !(found_identity_arm && found_diverging_arm) {
+                    return;
+                }
+                span_lint(cx, MANUAL_LET_ELSE, stmt.span, "this could be rewritten as `let else`");
```

Comment:
This lint should be machine applicable. Let's address my other review comments first and then we can look into how to get there.

Reply:
That's pretty hard because of the many challenges. E.g. you have to do pattern editing to end up with the identifiers the user originally wanted. As in, take this code:

```Rust
let foo = if let En::Hi { hi } = e { hi } else {...};
```

You have to convert it to:

```Rust
let En::Hi { hi: foo } = e else {...};
```

You also have to watch out for unused bindings. E.g. in this instance:

```Rust
let _bar = 0;
let e = (Some(0), 1);
let foo = if let (Some(hi), _bar) = e { hi } else {...};
println!("{_bar}");
```

It would print 0. You don't want it to do:

```Rust
let _bar = 0;
let e = (Some(0), 1);
let (Some(foo), _bar) = e else {...};
println!("{_bar}");
```

As this would print 1, as `_bar` is shadowed. You could change it to:

```Rust
let _bar = 0;
let e = (Some(0), 1);
let (Some(foo), _) = e else {...};
println!("{_bar}");
```

but that might not be wanted by users, as the `_bar` might still be useful for documentation purposes. Ideally that `_bar -> _` change would only happen if `_bar` would be shadowing an already existing binding.

I'm not saying it's impossible to make it machine applicable, all I'm saying is that it's a big task that I initially deferred for later.

Reply:
Good point, this lint can't really be machine applicable. Covering all corner cases here would be a nightmare. What do you think about suggesting 

```rust
let En::Hi { hi } = e else {...};
```

or

```rust
let (Some(hi), _bar) = e else {...};
```

And set `Applicability::HasPlaceholders`? That way rustfix won't auto-apply it, but we still got a suggestion. Maybe add an additional note, that variable names might have to be changed.